torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1341,6 +1343,274 @@ def __init__(self, var_to_sources):
+
+    def solve(self):
+        # as long as there are symbols with equalities, solve for them
+        # NOTE(avik): this is guaranteed to terminate (#iterations <= #symbols)
+        while(self._symbols_with_equalities):
+            s = self._symbols_with_equalities.pop()
+            exprs = self._univariate_inequalities.pop(s)
+            solution = sympy.solvers.inequalities.reduce_inequalities(exprs, s)
+            if isinstance(solution, sympy.And):
+                solution = next((arg for arg in solution.args if isinstance(arg, sympy.Eq)), solution)
+            assert isinstance(solution, sympy.Eq), f"Expected an equality constraint for {s}, got {solution}"
+            symbol, val = solution.args
+            assert symbol == s, f"Expected a constraint on {s} instead of on {symbol}"
+            # because this is univariate, the solution is a specialization
+            self._static_results.add(f"{self._dcp.symbol_to_source[s][0].name()} = {val}")
+            # add this as a substitution to simplify other constraints
+            self._substitutions[s] = val
+
+            # simplify multivariate inequalities: some of them will now become univariate!
+            multivariate_inequalities = self._multivariate_inequalities
+            self._multivariate_inequalities = set()
+            for expr in multivariate_inequalities:
+                self.add(expr.subs(s, self._substitutions[s]))
+
+            # simplify symbolic equivalences: some of them will now become specializations!
+            symbolic_equivalences = self._symbolic_equivalences
+            self._symbolic_equivalences = []
+            for source, expr in symbolic_equivalences:
+                self.add_equality(source, expr.subs(s, self._substitutions[s]))
+
+        # remaining symbols have only pure inequalities (no equalities)
+        for s, exprs in self._univariate_inequalities.items():
+            solution = sympy.solvers.inequalities.reduce_inequalities(exprs, s)
+            # because this is univariate, the solution is a dynamic (range) constraint
+            self._dynamic_results.add(self._dcp.doprint(solution))
+
+        # remaining symbolic equivalences become dynamic equality constraints
+        for source, expr in self._symbolic_equivalences:
+            self._dynamic_results.add(f"{self._dcp.print_source(source)} == {self._dcp.doprint(expr)}")
+
+        # solve linear congruences
+        # NOTE(avik): We do not need to solve them for symbols that have already been specialized.
+        reduced_congruences = self.reduce_congruences()
+        for s, congruences in reduced_congruences.items():
+            for congruence in congruences:
+                # any congruence that cannot be checked becomes a dynamic constraint as well
+                if not (s in self._substitutions) or not sympy.checksol(congruence, {s: self._substitutions[s]}):
+                    self._dynamic_results.add(self._dcp.doprint(sympy.Eq(congruence, 0)))
```

Comment:
Are you at all willing to write a translation validator for your solving here via Z3? Based on the discussion at https://github.com/pytorch/pytorch/pull/96616#issuecomment-1484839906

This would give a lot more confidence than just the tests.

Reply:
I'm up for it. The best I do here is ensure that at least the concrete values are solutions of original and simplified. But yeah, no equivalence / implication checks.